shared/eventbus/eventbus.go

```diff
@@ -147,10 +159,46 @@ func (b *BrokerService) Start() {
 }
 
 func (b *BrokerService) runHandlers() {
-	for {
-		ev := <-b.channel
+	for ev := range b.channel {
 		for _, subscriber := range b.subscribers[ev.Topic] {
-			go subscriber.handler(ev)
+			b.runGoroutine(func() { subscriber.handler(ev) })
 		}
 	}
 }
+
+// runGoroutine creates a goroutine, but maintains a record of it to ensure that execution completes before
+// the server is shutdown.
+func (b *BrokerService) runGoroutine(f func()) {
+	b.goroutineLimitChannel <- struct{}{}
+	atomic.AddInt32(&b.goroutineCount, 1)
+
+	go func() {
+		f()
+		<-b.goroutineLimitChannel
+
+		atomic.AddInt32(&b.goroutineCount, -1)
+		select {
+		case b.goroutineExitSignal <- struct{}{}:
+		default:
+		}
+	}()
+}
+
+// Stop blocks until all goroutines created by runGoroutine are finish or exists
+// when time is out.
+func (b *BrokerService) Stop() error {
+	done := make(chan struct{}, 1)
+	go func() {
+		for atomic.LoadInt32(&b.goroutineCount) != 0 {
+			<-b.goroutineExitSignal
+		}
+		done <- struct{}{}
+	}()
+
+	select {
+	case <-done:
+		return nil
```

Comment:
Sorry forgot the mention, we should also close the `b.channel` somewhere here to let that `runHandlers()` exit.

Reply:
Is this supposed to be fixed before the merge or just a comment?

Reply:
@iomodo I see you send struct into `b.stoppedChannel <- struct{}{}`, this indeed makes the following block in the `runHandlers()` to exit:
```Go
// checks whether the event bus was stopped
select {
case <-b.stoppedChannel:
	return
default:
}
```
Just wanted to discuss would it be better to close channel and take care on this on the run (ie. in the `Start` method)? Because, there is no guarantee that we run `Start` method only once, we should have a mechanism to be defensive on that. It's not necessarily related to this PR or needs to be addressed immediately. Something to discuss, in fact we can handle this within our team as well. WDYT @noxer?

Reply:
Couldn't the whole thing be much simpler with just using a `sync.WaitGroup` for those go routines?